internal/build/docker.go

```diff
@@ -189,79 +191,87 @@ func (l *localDockerBuilder) buildFromDfWithFiles(ctx context.Context, df Docker
 	return getDigestFromAux(*result)
 }
 
-func (l *localDockerBuilder) execStepsOnImage(ctx context.Context, img digest.Digest, steps []model.Cmd) (digest.Digest, error) {
-	imageWithSteps := img
-	for _, s := range steps {
-		cId, err := l.startContainer(ctx, containerConfigRunCmd(imageWithSteps, s))
-		if err != nil {
-			return "", fmt.Errorf("startContainer '%s': %v", img, err)
-		}
-
-		id, err := l.dcli.ContainerCommit(ctx, cId, types.ContainerCommitOptions{})
-		if err != nil {
-			return "", fmt.Errorf("containerCommit: %v", err)
-		}
-		imageWithSteps = digest.Digest(id.ID)
+func (l *localDockerBuilder) execInContainer(ctx context.Context, cID containerID, cmd model.Cmd) error {
+	created, err := l.dcli.ContainerExecCreate(ctx, string(cID), types.ExecConfig{
+		Cmd: cmd.Argv,
+	})
+	if err != nil {
+		return err
 	}
 
-	return digest.Digest(imageWithSteps), nil
-}
-
-// TODO(maia): can probs do this in a more efficient place -- e.g. `execStepsOnImage`
-// already spins up + commits a container, maybe piggyback off that?
-func (l *localDockerBuilder) imageWithEntrypoint(ctx context.Context, img digest.Digest, entrypoint model.Cmd) (digest.Digest, error) {
-	cId, err := l.startContainer(ctx, containerConfigRunCmd(img, model.Cmd{}))
-	if err != nil {
-		return "", fmt.Errorf("startContainer '%s': %v", string(img), err)
+	execID := execID(created.ID)
+	if execID == "" {
+		return fmt.Errorf("execInContainer: failed to create")
 	}
 
-	// Only attach the entrypoint if there's something in it
-	opts := types.ContainerCommitOptions{}
-	if !entrypoint.Empty() {
-		opts.Changes = []string{entrypoint.EntrypointStr()}
+	attached, err := l.dcli.ContainerExecAttach(ctx, string(execID), types.ExecStartCheck{})
+	if err != nil {
+		return fmt.Errorf("execInContainer#attach: %v", err)
 	}
+	defer attached.Close()
 
-	id, err := l.dcli.ContainerCommit(ctx, cId, opts)
+	// TODO(nick): feed this reader into the logger
+	buf := bytes.NewBuffer(nil)
+	_, err = io.Copy(buf, attached.Reader)
```

Comment:
i know at some point we discussed a way to read from this buf until we encountered EOF -- would that let us simplify the `inspect` loop below?

Reply:
i think you're confused. the existing code already reads from this buf until we encounter EOF. The inspect loop is a separate unrelated thing, because we have to wait until the exit code gets written.

Reply:
...yes, i was indeed confused. that makes sense, thanks!